src/Routes/RoutesFilter.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mezzio\Tooling\Routes;
+
+use ArrayIterator;
+use FilterIterator;
+use Mezzio\Router\Route;
+
+use function array_intersect;
+use function array_walk;
+use function get_class;
+use function in_array;
+use function is_array;
+use function is_string;
+use function preg_match;
+use function sprintf;
+use function str_replace;
+use function strtoupper;
+
+/**
+ * RoutesFilter filters a traversable list of Route objects based on any of the four Route criteria,
+ * those being the route's name, path, middleware, or supported method(s).
+ */
+class RoutesFilter extends FilterIterator
+{
+    /**
+     * An array storing the list of route options to filter a route on along with their respective values.
+     *
+     * The four allowed route options are: name, path, method, and middleware.
+     * Name and path can be a fixed string, such as user.profile, or a regular expression, such as user.*.
+     * Middleware can only contain a class name. Method can be either a string which contains one of the
+     * allowed HTTP methods, or an array of HTTP methods.
+     *
+     * @var array
+     */
+    private array $filterOptions;
+
+    public function __construct(ArrayIterator $routes, array $filterOptions = [])
+    {
+        parent::__construct($routes);
+
+        $this->filterOptions = $filterOptions;
+
+        // Filter out any options that are, effectively, "empty".
```

Comment:
You would not need to filter anything if you provide a proper option object. Can we probably change the code so we get rid of that array?